### ISSUE TITLE
CFA doesn't see non executable code

### ISSUE THREAD

**Issue by cevek** [Microsoft/TypeScript#15089]
**TypeScript Version:**  2.3.0-dev.20170403

**Code**

```ts
let validate = true;
if (validate) {

} else {
    // here validate type is never, but error doesn't arise
    if (validate) { 

    }
}

```

**Comment by aluanhaddad** [comment]
I remember this working previously. Perhaps a regression

**Comment by Andarist** [comment]
I don't think that the current intention is to report unreachable code if a value gets narrowed down to `never`. So at the very least, this should be re-labeled as Suggestion or something. cc @jakebailey

**Comment by RyanCavanaugh** [comment]
Yeah, I don't really understand what the proposed semantics are here. `Debug.assert(someNeverValue)` is a common pattern and it's not obvious what to do with a `never` value in an `if` either. People have argued that `neverValue.zzzzzqqqqwww` should be valid from theoretical grounds - truthiness testing is much closer to "always OK".
